pkg/cmd/experimental/syncgroups/rfc2307/ldapinterface.go

```diff
@@ -0,0 +1,154 @@
+package rfc2307
+
+import (
+	"github.com/go-ldap/ldap"
+
+	"github.com/openshift/origin/pkg/auth/ldaputil"
+)
+
+// NewLDAPInterface builds a new LDAPInterface using a schema-appropriate config
+func NewLDAPInterface(clientConfig ldaputil.LDAPClientConfig,
+	groupQuery ldaputil.IdentifiyingLDAPQueryOptions,
+	userQuery ldaputil.IdentifiyingLDAPQueryOptions,
+	groupMembershipAttributes []string) LDAPInterface {
+	return LDAPInterface{
+		clientConfig:              clientConfig,
+		groupQuery:                groupQuery,
+		userQuery:                 userQuery,
+		groupMembershipAttributes: groupMembershipAttributes,
+		cachedUsers:               make(map[string]*ldap.Entry),
+		cachedGroups:              make(map[string]*ldap.Entry),
+	}
+}
+
+// LDAPInterface extracts the member list of an LDAP group entry from an LDAP server
+// with first-class LDAP entries for groups. The LDAPInterface is *NOT* thread-safe.
+// The LDAPInterface satisfies:
+// - LDAPMemberExtractor
+// - LDAPGroupGetter
+// - LDAPGroupLister
+type LDAPInterface struct {
+	// clientConfig holds LDAP connection information
+	clientConfig ldaputil.LDAPClientConfig
+	// groupQuery holds the information necessary to make an LDAP query for a specific
+	// first-class group entry on the LDAP server
+	groupQuery ldaputil.IdentifiyingLDAPQueryOptions
+	// userQuery holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server
+	userQuery ldaputil.IdentifiyingLDAPQueryOptions
+	// groupMembershipAttributes defines which attributes on an LDAP user entry will be interpreted
```

Comment:
This comment isn't what I was expecting.  It suggests that the LDAP user entry has attributes that describe group membership. Wasn't this schema "first class groups, first class users, group membership on the groups"?  Also, clarify that in your comment on the type.